actions/rules/action_state_to_metastate.go

```diff
@@ -0,0 +1,467 @@
+		return t.Values, nil
+	}
+	return nil, errs.New("given field on workitemtype " + wit.ID.String() + " is not an enum field: " + fieldName)
+}
+
+// getStateToMetastateMap returns the mapping from state to metastate values read from the template.
+func (act ActionStateToMetaState) getStateToMetastateMap(workitemTypeID uuid.UUID) (map[string]string, error) {
+	wit, err := act.Db.WorkItemTypes().Load(act.Ctx, workitemTypeID)
+	if err != nil {
+		return nil, err
+	}
+	stateList, err := act.getValueListFromEnumField(wit, workitem.SystemState)
+	if err != nil {
+		return nil, err
+	}
+	metastateList, err := act.getValueListFromEnumField(wit, workitem.SystemMetaState)
+	if err != nil {
+		return nil, err
+	}
+	if len(stateList) != len(metastateList) {
+		return nil, errs.Errorf("inconsistent number of states and metatstates in the current work item type (must be equal): %d states != %d metastates", len(stateList), len(metastateList))
+	}
+	stateToMetastateMap := make(map[string]string)
+	for idx := range stateList {
+		thisState, ok := stateList[idx].(string)
+		if !ok {
+			return nil, errs.New("state value in value list is not of type string")
+		}
+		thisMetastate, ok := metastateList[idx].(string)
+		if !ok {
+			return nil, errs.New("metastate value in value list is not of type string")
+		}
+		// only the first state<->metatstate mapping needs to be taken as
+		// per definition. So we're only adding a new entry here if there
+		// is not an entry already existing.
+		if _, ok := stateToMetastateMap[thisState]; !ok {
+			stateToMetastateMap[thisState] = thisMetastate
+		}
+	}
+	return stateToMetastateMap, nil
+}
+
+// getMetastateToStateMap returns the mapping from metastate to state values read from the template.
+func (act ActionStateToMetaState) getMetastateToStateMap(workitemTypeID uuid.UUID) (map[string]string, error) {
+	stateToMetastate, err := act.getStateToMetastateMap(workitemTypeID)
+	if err != nil {
+		return nil, err
+	}
+	metastateToStateMap := make(map[string]string)
+	for state, metastate := range stateToMetastate {
+		// this is important: we only add a new entry if there
+		// is not an entry already existing. Therefore, we're only
+		// using the first mapping, satisfying the req for getting the
+		// first matching state for a metastate.
+		if _, ok := metastateToStateMap[metastate]; !ok {
+			metastateToStateMap[metastate] = state
+		}
+	}
+	return metastateToStateMap, nil
+}
+
+func (act ActionStateToMetaState) addOrUpdateChange(changes change.Set, attributeName string, oldValue interface{}, newValue interface{}) change.Set {
+	for _, change := range changes {
+		if change.AttributeName == attributeName {
+			change.NewValue = newValue
+			return changes
+		}
+	}
+	newChanges := append(changes, change.Change{
+		AttributeName: attributeName,
+		OldValue:      oldValue,
+		NewValue:      newValue,
+	})
+	return newChanges
+}
+
+func (act ActionStateToMetaState) fuseChanges(c1 change.Set, c2 change.Set) *change.Set {
```

Comment:
This function could need an explanation. And it I think it is best implemented as a member of `change.Set`, no? Then you can write:

```go
var a, b change.Set
c := a.fuse(b)
```

Also, why does it return a pointer to a `change.Set`? Below you explicitly dereference: `return newContext, *actionChanges, nil`.

Reply:
I agree `a.fuse(b)` makes more sense to me.

Reply:
No, because this implements a very specific functionality that is not generalizable. In this case, existing changes sould be updated rather than appended in order. This is very specific to this particular use case. I don't think this makes sense as part of `change.Set`. Also, I externalize to local functions for structuring the code and making is more understandable, not necessarily because there is a re-use case or it is generalizable functionality. It just does not make sense to write methods that have more than a few hundred lines. If it has, I segment them down into chunks that have a defined signature interface.

Reply:
And it returns `*actionChanges` because of exactly this reason. This is a structural method. The caller needs `*actionChanges`.

Reply:
That is somewhat confusing to me. The consumer of the `fuseChanges` function is `func (act ActionStateToMetaState) OnChange(newContext change.Detector, contextChanges change.Set, configuration string, actionChanges *change.Set) (change.Detector, change.Set, error)`. Inside of that function you assign the `actionChanges` like so: `actionChanges = act.fuseChanges(*actionChanges, executionChanges)` and then you also return the changes like so: `return newContext, *actionChanges, nil`. 

In the return you dereference the `actionChanges`, so for that part you don't need `fuseChanges` to return a pointer.

Since you also set a pointer that was a parameter before and also return the same value in `OnChange`, I doubt that this is thought through. One should be enough, either set (using the parameter) or return a new value but not do both.

Reply:
I understand the first point. I don't understand the second point. Anyways, I changed the return type fo `fuseChanges`. 5001aa2.